smart-home/src/XMLWriter.cpp

```diff
@@ -0,0 +1,68 @@
+#include "XMLWriter.h"
+
+#include "ACUnitDevice.h"
+#include "DoorbellDevice.h"
+#include "LightDevice.h"
+
+#include "TemperatureSensor.h"
+#include "HumiditySensor.h"
+#include "DistanceSensor.h"
+
+#include <algorithm>
+
+
+namespace smart_home_app {
+void XMLWriter::Write(std::string filename) {
+    pugi::xml_document doc;
+    pugi::xml_node root = doc.append_child("Root");
+
+    for (Room r : rooms_) {
+        pugi::xml_node room_root = root.append_child("Room");
+        room_root.append_attribute("name").set_value(r.GetRoomName().c_str());
+
+        for (Device* d : r.GetDevices()) {
+            pugi::xml_node device_root = room_root.append_child("Device");
```

Comment:
I understand, now I am just throwing an exception in case of an error if the pointer is nullptr, so that I would not go about trying to write a Device@nullptr.